### PR TITLE
Ravendb-18554 - Queries should failover if the index is compacting.

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18554

### Additional description

Fixing - Queries should failover if the index is compacting.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                else
+                {
+                    Assert.Null(e); // Faliover
+                }
+            }
+        }
+
+        class Categoroies_Details : AbstractMultiMapIndexCreationTask<Categoroies_Details.Entitiy>
+        {
+            internal class Entitiy
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]:
```suggestion
            internal class Entity
```

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            {
+                public string Id { get; set; }
+                public string Details { get; set; }
+
+                public override string ToString()
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]: Not needed

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                    catch (Exception e1)
+                    {
+                        e = e1;
+                    }
+                };
+
+                var operation = await store.Maintenance.Server.SendAsync(new CompactDatabaseOperation(settings));
+                await operation.WaitForCompletionAsync();
+
+                if (!cluster)
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]:
```suggestion
                if (cluster == false)
```

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                    {
+                        e = e1;
+                    }
+                };
+
+                var operation = await store.Maintenance.Server.SendAsync(new CompactDatabaseOperation(settings));
+                await operation.WaitForCompletionAsync();
+
+
+                if (!cluster)
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]:
```suggestion
                if (cluster == false)
```

File: src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
```
Comment (grisha-kotler) [inline src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs]: Do we need this exception in the `Client`?

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 6]
```diff
@@ -1771,6 +1771,10 @@ internal class TestingStuff
 
             internal Action CollectionRunnerBeforeOpenReadTransaction;
 
+            internal Action CompactionAfterDatabaseUnloadAction;
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DocumentDatabase.cs]:
```suggestion
            internal Action CompactionAfterDatabaseUnload;
```

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 7]
```diff
@@ -4578,6 +4579,8 @@ public void Compact(Action<IOperationProgress> onProgress, CompactionResult resu
                 PathSetting compactPath = null;
                 PathSetting tempPath = null;
 
+                DocumentDatabase.ForTestingPurposes?.IndexCompaction?.Invoke();
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/Index.cs]: Use the `ForTestingPurposes` of the `IndexStore`

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+                // Test
+                Exception e = null;
+                database.ForTestingPurposesOnly().CompactionAfterDatabaseUnloadAction = () =>
+                {
+                    try
+                    {
+                        using (var session = store.OpenSession())
+                        {
+                            session.Query<Categoroies_Details.Entitiy, Categoroies_Details>()
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]: Check that we got the correct result here (for cluster)

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+                // Test
+                Exception e = null;
+                database.ForTestingPurposesOnly().IndexCompaction = () =>
+                {
+                    try
+                    {
+                        using (var session = store.OpenSession())
+                        {
+                            session.Query<Categoroies_Details.Entitiy, Categoroies_Details>()
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]: Check that we got the correct result here (for cluster)

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs:1]: Yes

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 7 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Index.cs]: And shouldn't it be after `RestartEnvironment`?

File: test/SlowTests/Issues/RavenDB-18554.cs (hunk 1 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18554.cs]: Done

File: test/SlowTests/Issues/RavenDB-18554.cs (hunk 2 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18554.cs]: Done

File: test/SlowTests/Issues/RavenDB-18554.cs (hunk 3 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18554.cs]: Done

File: test/SlowTests/Issues/RavenDB-18554.cs (hunk 4 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18554.cs]: Done

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 6 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: Done

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 7 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Done

File: test/SlowTests/Issues/RavenDB-18554.cs (hunk 8 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18554.cs]: Done

File: test/SlowTests/Issues/RavenDB-18554.cs (hunk 9 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18554.cs]: Done

File: src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs]: Yes, Do you want me to use another exception instead?

Comment (grisha-kotler) [reply src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs]: No. Just moved to the server.

Comment (grisha-kotler) [reply src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs]: What about this?

File: test/SlowTests/Issues/RavenDB-18554.cs [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                    }
+                    catch (Exception e)
+                    {
+                        exception = e;
+                    }
+                };
+
+                if (cluster)
+                {
+                    foreach (var node in nodes)
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18554.cs]: You shouldn't set the task for all nodes in the cluster.
Only the relevant node.

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Raven.Client.Exceptions.Database
+{
+    internal class IndexCompactionInProgressException : RavenException
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs]: move to Raven.Client.Exceptions.Documents.Indexes

Comment (shaharhikri) [reply src/Raven.Client/Exceptions/Database/IndexCompactionInProgressException.cs]: Done

**Review by grisha-kotler — APPROVED** [review]

**Pull request merged**